martin/src/cog/mod.rs

```diff
@@ -0,0 +1,383 @@
+    #[allow(clippy::cast_sign_loss)]
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::too_many_lines)]
+    async fn get_tile(
+        &self,
+        xyz: TileCoord,
+        _url_query: Option<&UrlQuery>,
+    ) -> MartinResult<TileData> {
+        let tif_file =
+            File::open(&self.path).map_err(|e| FileError::IoError(e, self.path.clone()))?;
+        let mut decoder =
+            Decoder::new(tif_file).map_err(|e| CogError::InvalidTifFile(e, self.path.clone()))?;
+        decoder = decoder.with_limits(tiff::decoder::Limits::unlimited());
+
+        let ifd = self.meta.zoom_and_ifd.get(&(xyz.z)).ok_or_else(|| {
+            CogError::ZoomOutOfRange(
+                xyz.z,
+                self.path.clone(),
+                self.meta.min_zoom,
+                self.meta.max_zoom,
+            )
+        })?;
+
+        decoder
+            .seek_to_image(*ifd)
+            .map_err(|e| CogError::IfdSeekFailed(e, *ifd, self.path.clone()))?;
+
+        let tiles_across = self
+            .meta
+            .zoom_and_tile_across_down
+            .get(&(xyz.z))
+            .ok_or_else(|| {
+                CogError::ZoomOutOfRange(
+                    xyz.z,
+                    self.path.clone(),
+                    self.meta.min_zoom,
+                    self.meta.max_zoom,
+                )
+            })?
+            .0;
+        let tile_idx = xyz.y * tiles_across + xyz.x;
+        let decode_result = decoder
+            .read_chunk(tile_idx)
+            .map_err(|e| CogError::ReadChunkFailed(e, tile_idx, *ifd, self.path.clone()))?;
+        let color_type = decoder
+            .colortype()
+            .map_err(|e| CogError::InvalidTifFile(e, self.path.clone()))?;
+
+        let tile_width = decoder.chunk_dimensions().0;
+        let tile_height = decoder.chunk_dimensions().1;
+        let (data_width, data_height) = decoder.chunk_data_dimensions(tile_idx);
+
+        //do more research on the not u8 case, is this the right way to do it?
+        let png_file_bytes = match (decode_result, color_type) {
+            (DecodingResult::U8(vec), tiff::ColorType::RGB(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                3,
+                self.meta.nodata.map(|v| v as u8),
+                &self.path,
+            ),
+            (DecodingResult::U8(vec), tiff::ColorType::RGBA(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                4,
+                self.meta.nodata.map(|v| v as u8),
+                &self.path,
+            ),
+            (_, _) => Err(CogError::NotSupportedColorTypeAndBitDepth(
+                color_type,
+                self.path.clone(),
+            )),
+            // do others in next PRs, a lot of disscussion would be needed
+        }?;
+        Ok(png_file_bytes)
+    }
+}
+
+fn rgb_to_png(
```

Comment:
Tests from test.sh for this would be better I guess as this part is fragile for upcoming PRs. Let's do tests from test.sh like we don't know the implementation details.